### PR TITLE
Chore/preserve old img hashes nf

### PR DESCRIPTION
**Pull request by nss10**
Link to JIRA ticket if there is one:  [MIDRC-766](https://ctds-planx.atlassian.net/browse/MIDRC-766)

relates to https://github.com/uc-cdis/midrc-scripts/pull/17

### New Features
* GH workflow to generate image hashes now supports historical values upto 2 (configurable) instances

### Breaking Changes
* Corresponding logic at X-soar's script [here](https://github.com/uc-cdis/midrc-scripts/blob/e0a8fce2e77adb70fd3aa3b00cfd9322bc7c64f6/container_evaluation_scripts/cortex_xsoar/container_vulnerability_checker.py#L110C1-L110C55) must be updated to go hand in hand

[MIDRC-766]: https://ctds-planx.atlassian.net/browse/MIDRC-766?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 1]
```diff
@@ -18,27 +18,78 @@ jobs:
 
       - name: Pull HTTP Response
         run: |
+          echo "Line $LINENO: starting the script"
           url_list=$(curl -s "https://raw.githubusercontent.com/uc-cdis/containers/master/nextflow-base-images/allowed_base_images.txt")
+          max_version_count=2  # Configurable value which indicates maximum of how many versions of past image hashes do we support
           TOKEN=$(curl -s https://public.ecr.aws/token/ | jq -r .token)
-          layer_json="{}"
+          output_file_name="nextflow-base-images/approved_img_layers_2.json" # change it once the xsoar script is ready to read the current file.
```
Comment (nss10) [inline .github/workflows/get_layer_info_for_nf_imgs.yml:25]: Do not merge PR unless this is changed.

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 2]
```diff
@@ -18,27 +18,78 @@ jobs:
...
+
+            if [ $? -ne 0 ] || [ -z "$response" ]; then
+              echo "Line $LINENO: Failed to retrieve valid response from manifest_url -- $manifest_url" >&2
+              continue
+            fi
+
+            current_value=$(echo "$layer_json" | jq -e ".[\"$image_url\"]" 1>/dev/null || echo "{}")
+            current_value=$(echo "$current_value" | jq --argjson k "$max_version_count" 'to_entries | sort_by(.key) | reverse | if length > $k then .[:$k] else . end | from_entries')
+            value=$(echo "$current_value" | jq --arg key "$(date "+%Y_%m_%d_%H:%M:%S")" --argjson response "$response" '. + { ($key) : $response }' | jq 'to_entries | sort_by(.key) | reverse | from_entries')
+            layer_json=$(echo "$layer_json" | jq --arg key "$image_url" --argjson value "$value" '. + { ($key): $value }')
```
Comment (paulineribeyre) [inline .github/workflows/get_layer_info_for_nf_imgs.yml:86]: could you add comments for these lines, they are very dense to read 😅

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 3]
```diff
@@ -18,27 +18,78 @@ jobs:
...
+            # Fetch the manifest with retry using exponential backoff
+            response=$(fetch_manifest "$manifest_url" | jq "[.layers[].digest]|.[-1]")
+
+            if [ $? -ne 0 ] || [ -z "$response" ]; then
+              echo "Line $LINENO: Failed to retrieve valid response from manifest_url -- $manifest_url" >&2
+              continue
+            fi
+
+            current_value=$(echo "$layer_json" | jq -e ".[\"$image_url\"]" 1>/dev/null || echo "{}")
+            current_value=$(echo "$current_value" | jq --argjson k "$max_version_count" 'to_entries | sort_by(.key) | reverse | if length > $k then .[:$k] else . end | from_entries')
```
Comment (paulineribeyre) [inline .github/workflows/get_layer_info_for_nf_imgs.yml:84]: it looks like you got this working for the case where the base image does not exist in the file yet, but can you confirm that it's been tested?

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 4]
```diff
@@ -18,27 +18,78 @@ jobs:
...
+            value=$(echo "$current_value" | jq --arg key "$(date "+%Y_%m_%d_%H:%M:%S")" --argjson response "$response" '. + { ($key) : $response }' | jq 'to_entries | sort_by(.key) | reverse | from_entries')
+            layer_json=$(echo "$layer_json" | jq --arg key "$image_url" --argjson value "$value" '. + { ($key): $value }')
+
           done <<< "$url_list"
-          echo "$layer_json" > nextflow-base-images/approved_img_layers.json
+
+          if [ -n "$layer_json" ]; then
+            echo "$layer_json" >  $output_file_name
+          fi
```
Comment (paulineribeyre) [inline .github/workflows/get_layer_info_for_nf_imgs.yml:92]: in what case would we not enter this block?

File: .github/workflows/get_layer_info_for_nf_imgs.yml (hunk 1 above)
Comment (paulineribeyre) [reply .github/workflows/get_layer_info_for_nf_imgs.yml:25]: also need to revert `# branches: master`

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 5]
```diff
@@ -18,27 +18,78 @@ jobs:
...
+            for ((i=0; i<retries; i++)); do
+              response=$(curl -s -H "Authorization: Bearer $TOKEN" "$url")
+              error_code=$(echo "$response" | jq -r '.errors[0].code // empty')
+
+              if [ "$error_code" == "TOOMANYREQUESTS" ]; then
+                echo "Line $LINENO: Rate limit exceeded, waiting for $wait_time seconds before retrying..." >&2
+                sleep $wait_time
+                wait_time=$((wait_time * 2))  # Exponential backoff
+              else
+                echo "$response"
```
[the remark this answers is not in the document]
Comment (nss10) [reply .github/workflows/get_layer_info_for_nf_imgs.yml:50]: Any echo statement without specifically mentioning an output stream, it acts like returning a value from the function.

File: .github/workflows/get_layer_info_for_nf_imgs.yml (hunk 3 above)
Comment (nss10) [reply .github/workflows/get_layer_info_for_nf_imgs.yml:84]: Oh yes, I did test this with the following scenarios

If the old file has no key with the image name,
If the old file is missing altogether
If the old file has just one value with the date and we need to add more values
If the old file has several values(say 4) and the new N-k value has k=2 and we need to reduce number of old values.

File: .github/workflows/get_layer_info_for_nf_imgs.yml (hunk 4 above)
Comment (nss10) [reply .github/workflows/get_layer_info_for_nf_imgs.yml:92]: if there are any errors causing layers_json to return an empty string, this `if condition` prevents us from overwriting the existing layers file.

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 6]
```diff
@@ -18,27 +18,78 @@ jobs:
 
       - name: Pull HTTP Response
         run: |
+          echo "Line $LINENO: starting the script"
           url_list=$(curl -s "https://raw.githubusercontent.com/uc-cdis/containers/master/nextflow-base-images/allowed_base_images.txt")
+          max_version_count=2  # Configurable value which indicates maximum of how many versions of past image hashes do we support
           TOKEN=$(curl -s https://public.ecr.aws/token/ | jq -r .token)
-          layer_json="{}"
+          output_file_name="nextflow-base-images/approved_img_layers_2.json" # change it once the xsoar script is ready to read the current file.
+          echo "Line $LINENO: After setting up basic variables"
```
[the remark this answers is not in the document]
Comment (nss10) [reply .github/workflows/get_layer_info_for_nf_imgs.yml:26]: No, taking them off.

File: .github/workflows/get_layer_info_for_nf_imgs.yml [hunk 7]
```diff
@@ -16,29 +16,86 @@ jobs:
...
+            if [ $? -ne 0 ] || [ -z "$response" ]; then
+              echo "Line $LINENO: Failed to retrieve valid response from manifest_url -- $manifest_url" >&2
+              continue
+            fi
+
+            # Extracting current value of `image_url` if the key exists, otherwise return `{}`
+            current_value=$(echo "$layer_json" | jq -e ".[\"$image_url\"] // empty" || echo "{}")
+
+            # Eliminating older values if there are more than `max_version_count` and return all the remaining values.
+            current_value=$(echo "$current_value" | jq --argjson k "$max_version_count" 'to_entries | sort_by(.key) | reverse | if length > $k then .[:$k] else . end | from_entries')
```
Comment (paulineribeyre) [inline .github/workflows/get_layer_info_for_nf_imgs.yml:86]: this step doesn't use the timestamps to remove old values, it assumes the entries are in descending order, right? could you mention that in the comment

**Review by paulineribeyre — COMMENTED** [review]
looks good

File: .github/workflows/get_layer_info_for_nf_imgs.yml (hunk 7 above)
Comment (nss10) [reply .github/workflows/get_layer_info_for_nf_imgs.yml:86]: It doesn't assume the fields to be in desc order, but it actually sorts the fields in descending order, not based on timestamps but using string comparison.

**Review by paulineribeyre — APPROVED** [review]

**Pull request merged**